Mathlib/Topology/Category/CompHaus/EffectiveEpi.lean

```diff
@@ -0,0 +1,216 @@
+  ext ⟨_, h⟩ ; exact h
+
+def QB : CompHaus.{u} :=
+  haveI : T2Space (Quotient <| relation π) :=
+    ⟨fun _ _ h => separated_by_continuous (ι_fun_continuous π) <| (ι_fun_injective π).ne h ⟩
+  CompHaus.of (Quotient <| relation π)
+
+def ι_hom : (QB π) ⟶ B := ⟨ι_fun π, ι_fun_continuous π⟩
+
+noncomputable
+def ι : (QB π) ≅ B :=
+  haveI : IsIso (ι_hom π) := by
+    apply isIso_of_bijective
+    refine ⟨ι_fun_injective _, ?_⟩
+    intro b
+    obtain ⟨a,x,h⟩ := surj b
+    refine ⟨Quotient.mk _ ⟨a,x⟩, h⟩
+  asIso (ι_hom π)
+
+def π' : (a : α) → (X a ⟶ QB π) := fun a =>
+  { toFun := fun x => Quotient.mk _ ⟨a, x⟩
+    continuous_toFun := by
+      apply Continuous.comp
+      apply continuous_quot_mk
+      apply continuous_sigmaMk (σ := fun a => X a) }
+
+def struct_aux : EffectiveEpiFamilyStruct X (π' π) where
+  desc := fun {W} e h => {
+    toFun := Quotient.lift (fun ⟨a,x⟩ => e a x) <| by
+      rintro ⟨a,x⟩ ⟨b,y⟩ ⟨Z,z,fst,snd,hh,hx,hy⟩ ; dsimp at *
+      rw [← hx, ← hy]
+      specialize h _ _ fst snd ?_
+      · ext z
+        apply ι_fun_injective
+        apply_fun (fun q => q z) at hh
+        exact hh
+      apply_fun (fun q => q z) at h
+      exact h
+    continuous_toFun := by
+      apply Continuous.quotient_lift
+      apply continuous_sigma
+      intro a
+      exact (e a).continuous }
+  fac := by intro Z e h a ; ext ; rfl
+  uniq := by
+    intro Z e h m hm
+    ext ⟨⟨a,x⟩⟩
+    specialize hm a
+    apply_fun (fun q => q x) at hm
+    exact hm
+
+@[reassoc (attr := simp)]
+lemma π'_comp_ι_hom (a : α) : π' π a ≫ (ι _ surj).hom = π a := by ext ; rfl
+
+@[reassoc (attr := simp)]
+lemma π_comp_ι_inv (a : α) : π a ≫ (ι _ surj).inv = π' π a :=  by
+  rw [Iso.comp_inv_eq]
+  exact π'_comp_ι_hom _ surj _
+
+-- TODO: Make a general construction for transferring such structs along isomorphisms.
+noncomputable
+def struct : EffectiveEpiFamilyStruct X π where
+  desc := fun {W} e h => (ι π surj).inv ≫ (struct_aux π).desc e (fun {Z} a₁ a₂ g₁ g₂ hh => by
+      apply h
+      rw [← cancel_mono (ι _ surj).inv]
+      simpa only [Category.assoc, π_comp_ι_inv])
+  fac := by
+    intro W e h a
+    simp only [Eq.ndrec, id_eq, eq_mpr_eq_cast, π_comp_ι_inv_assoc, (struct_aux π).fac]
+  uniq := by
+    intro W e h m hm
+    dsimp
+    rw [Iso.eq_inv_comp]
+    apply (struct_aux π).uniq
+    intro a
+    simpa using hm a
+
+end EffectiveEpiFamily
+
+theorem effectiveEpiFamily_of_jointly_surjective
+    {α : Type} [Fintype α] {B : CompHaus.{u}}
+    (X : α → CompHaus.{u}) (π : (a : α) → (X a ⟶ B))
+    (surj : ∀ b : B, ∃ (a : α) (x : X a), π a x = b) :
+    EffectiveEpiFamily X π :=
+  ⟨⟨CompHaus.EffectiveEpiFamily.struct π surj⟩⟩
+
+open EffectiveEpiFamily
+
+theorem effectiveEpiFamily_tfae
+    {α : Type} [Fintype α] {B : CompHaus.{u}}
+    (X : α → CompHaus.{u}) (π : (a : α) → (X a ⟶ B)) :
+    [ EffectiveEpiFamily X π
+    , Epi (Sigma.desc π)
+    , ∀ b : B, ∃ (a : α) (x : X a), π a x = b
+    ].TFAE := by
```

Comment:
Suggest writing `List.TFAE`, rather than using dot notation, for readability!

Reply:
Good idea! Done.